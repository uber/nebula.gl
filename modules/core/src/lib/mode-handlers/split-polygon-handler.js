// @flow

import booleanPointInPolygon from '@turf/boolean-point-in-polygon';
import turfDifference from '@turf/difference';
import turfBuffer from '@turf/buffer';
import lineIntersect from '@turf/line-intersect';
import type { ClickEvent, PointerMoveEvent } from '../event-types.js';
import type { EditAction } from './mode-handler.js';
import { ModeHandler } from './mode-handler.js';

export class SplitPolygonHandler extends ModeHandler {
  handleClick(event: ClickEvent): ?EditAction {
    super.handleClick(event);

    const editAction: ?EditAction = null;
    const tentativeFeature = this.getTentativeFeature();
    const selectedGeometry = this.getSelectedGeometry();
    const clickSequence = this.getClickSequence();

    if (!selectedGeometry) {
      // eslint-disable-next-line no-console,no-undef
      console.warn('A polygon must be selected for splitting');
      this._setTentativeFeature(null);
      return editAction;
    }
    const pt = {
      type: 'Point',
      coordinates: event.groundCoords
    };
    const isPointInPolygon = booleanPointInPolygon(pt, selectedGeometry);
    if (clickSequence.length > 1 && tentativeFeature && !isPointInPolygon) {
      this.resetClickSequence();
      const isLineInterectingWithPolygon = lineIntersect(tentativeFeature, selectedGeometry);
      if (isLineInterectingWithPolygon.features.length === 0) {
        this._setTentativeFeature(null);
        return editAction;
      }
      return this.splitPolygon();
    }

    return editAction;
  }

  handlePointerMove({
    groundCoords
  }: PointerMoveEvent): { editAction: ?EditAction, cancelMapPan: boolean } {
    const clickSequence = this.getClickSequence();
    const result = { editAction: null, cancelMapPan: false };

    if (clickSequence.length === 0) {
      // nothing to do yet
      return result;
    }

    this._setTentativeFeature({
      type: 'Feature',
      geometry: {
        type: 'LineString',
        coordinates: [...clickSequence, groundCoords]
      }
    });

    return result;
  }

  splitPolygon() {
    const selectedGeometry = this.getSelectedGeometry();
    const tentativeFeature = this.getTentativeFeature();
    const featureIndex = this.getSelectedFeatureIndexes()[0];
    const modeConfig = this.getModeConfig() || {};

    // Default gap in between the polygon
    let { gap = 0.1, units = 'centimeters' } = modeConfig;
    if (gap === 0) {
      gap = 0.1;
      units = 'centimeters';
    }

    const buffer = turfBuffer(tentativeFeature, gap, { units });
    const updatedGeometry = turfDifference(selectedGeometry, buffer);
    this._setTentativeFeature(null);
    if (!updatedGeometry) {
      // eslint-disable-next-line no-console,no-undef
      console.warn('Canceling edit. Split Polygon erased');
      return null;
    }

    const { type, coordinates } = updatedGeometry.geometry;
    let updatedCoordinates = [];
    if (type === 'Polygon') {
      // Update the coordinates as per Multipolygon
      updatedCoordinates = coordinates.map(c => [c]);
    } else {
      // Handle Case when Multipolygon has holes
      updatedCoordinates = coordinates.reduce((agg, prev) => {
        prev.forEach(p => {
          agg.push([p]);
        });
        return agg;
      }, []);
    }

    // Update the type to Mulitpolygon
    const updatedData = this.getImmutableFeatureCollection().replaceGeometry(featureIndex, {
      type: 'MultiPolygon',
      coordinates: updatedCoordinates
    });

    const editAction: EditAction = {
      updatedData: updatedData.getObject(),
      editType: 'split',
      featureIndex,
      positionIndexes: null,
      position: null
    };

    return editAction;
  }
}
