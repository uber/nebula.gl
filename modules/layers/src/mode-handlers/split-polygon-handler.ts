import booleanPointInPolygon from '@turf/boolean-point-in-polygon';
import turfDifference from '@turf/difference';
import turfBuffer from '@turf/buffer';
import lineIntersect from '@turf/line-intersect';
import { lineString } from '@turf/helpers';
import turfBearing from '@turf/bearing';
import turfDistance from '@turf/distance';
import turfDestination from '@turf/destination';
import turfPolygonToLine from '@turf/polygon-to-line';
import nearestPointOnLine from '@turf/nearest-point-on-line';
import { generatePointsParallelToLinePoints } from '../utils';
import { ClickEvent, PointerMoveEvent } from '../event-types';
import { EditAction, ModeHandler } from './mode-handler';

// TODO edit-modes: delete handlers once EditMode fully implemented
export class SplitPolygonHandler extends ModeHandler {
  calculateGroundCoords(clickSequence: any, groundCoords: any) {
    const modeConfig = this.getModeConfig();
    if (!modeConfig || !modeConfig.lock90Degree || !clickSequence.length) {
      return groundCoords;
    }
    if (clickSequence.length === 1) {
      // if first point is clicked, then find closest polygon point and build ~90deg vector
      const firstPoint = clickSequence[0];
      const selectedGeometry = this.getSelectedGeometry();
      // @ts-ignore
      const feature = turfPolygonToLine(selectedGeometry);

      const lines = feature.type === 'FeatureCollection' ? feature.features : [feature];
      let minDistance = Number.MAX_SAFE_INTEGER;
      let closestPoint = null;
      // If Multipolygon, then we should find nearest polygon line and stick split to it.
      lines.forEach((line) => {
        const snapPoint = nearestPointOnLine(line, firstPoint);
        const distanceFromOrigin = turfDistance(snapPoint, firstPoint);
        if (minDistance > distanceFromOrigin) {
          minDistance = distanceFromOrigin;
          closestPoint = snapPoint;
        }
      });

      if (closestPoint) {
        // closest point is used as 90degree entry to the polygon
        const lastBearing = turfBearing(firstPoint, closestPoint);
        const currentDistance = turfDistance(firstPoint, groundCoords, { units: 'meters' });
        return turfDestination(firstPoint, currentDistance, lastBearing, {
          units: 'meters',
        }).geometry.coordinates;
      }
      return groundCoords;
    }
    // Allow only 90 degree turns
    const lastPoint = clickSequence[clickSequence.length - 1];
    const [approximatePoint] = generatePointsParallelToLinePoints(
      clickSequence[clickSequence.length - 2],
      lastPoint,
      groundCoords
    );
    // align point with current ground
    const nearestPt = nearestPointOnLine(lineString([lastPoint, approximatePoint]), groundCoords)
      .geometry.coordinates;
    return nearestPt;
  }

  handleClick(event: ClickEvent): EditAction | null | undefined {
    super.handleClick({
      ...event,
      groundCoords: this.calculateGroundCoords(this.getClickSequence(), event.groundCoords),
    });
    const editAction: EditAction | null | undefined = null;
    const tentativeFeature = this.getTentativeFeature();
    const selectedGeometry = this.getSelectedGeometry();
    const clickSequence = this.getClickSequence();

    if (!selectedGeometry) {
      // eslint-disable-next-line no-console,no-undef
      console.warn('A polygon must be selected for splitting');
      this._setTentativeFeature(null);
      return editAction;
    }
    const pt = {
      type: 'Point',
      coordinates: clickSequence[clickSequence.length - 1],
    };
    // @ts-ignore
    const isPointInPolygon = booleanPointInPolygon(pt, selectedGeometry);
    if (clickSequence.length > 1 && tentativeFeature && !isPointInPolygon) {
      this.resetClickSequence();
      // @ts-ignore
      const isLineInterectingWithPolygon = lineIntersect(tentativeFeature, selectedGeometry);
      if (isLineInterectingWithPolygon.features.length === 0) {
        this._setTentativeFeature(null);
        return editAction;
      }
      return this.splitPolygon();
    }

    return editAction;
  }

  handlePointerMove({
    groundCoords,
  }: PointerMoveEvent): { editAction: EditAction | null | undefined; cancelMapPan: boolean } {
    const clickSequence = this.getClickSequence();
    const result = { editAction: null, cancelMapPan: false };

    if (clickSequence.length === 0) {
      // nothing to do yet
      return result;
    }

    this._setTentativeFeature({
      type: 'Feature',
      geometry: {
        type: 'LineString',
        coordinates: [...clickSequence, this.calculateGroundCoords(clickSequence, groundCoords)],
      },
    });

    return result;
  }

  splitPolygon() {
    const selectedGeometry = this.getSelectedGeometry();
    const tentativeFeature = this.getTentativeFeature();
    const featureIndex = this.getSelectedFeatureIndexes()[0];
    const modeConfig = this.getModeConfig() || {};

    // Default gap in between the polygon
    let { gap = 0.1, units = 'centimeters' } = modeConfig;
    if (gap === 0) {
      gap = 0.1;
      units = 'centimeters';
    }
    // @ts-ignore
    const buffer = turfBuffer(tentativeFeature, gap, { units });
    // @ts-ignore
    const updatedGeometry = turfDifference(selectedGeometry, buffer);
    this._setTentativeFeature(null);
    if (!updatedGeometry) {
      // eslint-disable-next-line no-console,no-undef
      console.warn('Canceling edit. Split Polygon erased');
      return null;
    }

    const { type, coordinates } = updatedGeometry.geometry;
    let updatedCoordinates = [];
    if (type === 'Polygon') {
      // Update the coordinates as per Multipolygon
      // @ts-ignore
      updatedCoordinates = coordinates.map((c) => [c]);
    } else {
      // Handle Case when Multipolygon has holes
      // @ts-ignore
      updatedCoordinates = coordinates.reduce((agg, prev) => {
        prev.forEach((p) => {
          agg.push([p]);
        });
        return agg;
      }, []);
    }

    // Update the type to Mulitpolygon
    const updatedData = this.getImmutableFeatureCollection().replaceGeometry(featureIndex, {
      type: 'MultiPolygon',
      coordinates: updatedCoordinates,
    });

    const editAction: EditAction = {
      updatedData: updatedData.getObject(),
      editType: 'split',
      featureIndexes: [featureIndex],
      editContext: null,
    };

    return editAction;
  }
}
