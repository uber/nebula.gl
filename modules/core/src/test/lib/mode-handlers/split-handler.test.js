// @flow
/* eslint-env jest */
import { SplitPolygonHandler } from '../../../lib/mode-handlers/split-polygon-handler.js';
import type { EditAction } from '../../../lib/mode-handlers/mode-handler.js';
import {
  createFeatureCollection,
  createMultiPolygonFeature,
  createPolygonFeature,
  createClickEvent,
  createPointerMoveEvent
} from '../test-utils.js';

describe('Split Polygon Handler ', () => {
  test('sets tentative feature to a LineString after first click on Pointer move', () => {
    const featureCollection = createFeatureCollection();
    const handler = new SplitPolygonHandler(featureCollection);
    handler.setSelectedFeatureIndexes([2]);
    handler.handleClick(createClickEvent([1, 2]));
    handler.handlePointerMove(createPointerMoveEvent([2, 3]));
    const tentativeFeature = handler.getTentativeFeature();
    expect(tentativeFeature).toEqual({
      type: 'Feature',
      geometry: {
        type: 'LineString',
        coordinates: [[1, 2], [2, 3]]
      }
    });
  });

  describe('should reset tentativeFeature on handleClick ', () => {
    test('when selectedGeometry is not selected ', () => {
      const featureCollection = createFeatureCollection();
      const handler = new SplitPolygonHandler(featureCollection);
      const action = handler.handleClick(createClickEvent([1, 2]));
      const tentativeFeature = handler.getTentativeFeature();
      expect(action).toBeNull();
      expect(tentativeFeature).toBeNull();
    });

    test('when clicked point is inside the polygon ', () => {
      const featureCollection = createFeatureCollection();
      const handler = new SplitPolygonHandler(featureCollection);
      handler.setSelectedFeatureIndexes([2]);
      const action1 = handler.handleClick(createClickEvent([-2, -2]));
      handler.handlePointerMove(createPointerMoveEvent([-1, -1]));
      const action = handler.handleClick(createClickEvent([-1, -1]));
      expect(action1).toBeNull();
      expect(action).toBeNull();
    });

    test('when lineString is outside the polygon ', () => {
      const featureCollection = createFeatureCollection();
      const handler = new SplitPolygonHandler(featureCollection);
      handler.setSelectedFeatureIndexes([2]);
      const action1 = handler.handleClick(createClickEvent([-2, -2]));
      handler.handlePointerMove(createPointerMoveEvent([-4, -4]));
      const action = handler.handleClick(createClickEvent([-4, -4]));
      expect(action1).toBeNull();
      expect(action).toBeNull();
    });

    test('when lineString is outside the polygon lock90Degree', () => {
      const featureCollection = createFeatureCollection();
      const handler = new SplitPolygonHandler(featureCollection);
      handler.setModeConfig({ lock90Degree: true });
      handler.setSelectedFeatureIndexes([2]);
      const action1 = handler.handleClick(createClickEvent([-2, -2]));
      handler.handlePointerMove(createPointerMoveEvent([-4, -4]));
      const action = handler.handleClick(createClickEvent([-4, -4]));
      expect(action1).toBeNull();
      expect(action).toMatchSnapshot();
    });
  });

  test('should split the polygon with holes and upgrade to MultiPolygon ', () => {
    const featureCollection = {
      type: 'FeatureCollection',
      features: [createPolygonFeature()]
    };
    const handler = new SplitPolygonHandler(featureCollection);
    handler.setModeConfig({});
    handler.setSelectedFeatureIndexes([0]);
    handler.handleClick(createClickEvent([-2, -2]));
    handler.handlePointerMove(createPointerMoveEvent([2, 3]));
    const action = handler.handleClick(createClickEvent([2, 3]));

    const expectedAction: EditAction = {
      updatedData: {
        type: 'FeatureCollection',
        features: [
          {
            type: 'Feature',
            properties: {},
            geometry: {
              type: 'MultiPolygon',
              coordinates: [
                [
                  [
                    [-1, -1],
                    [1, -1],
                    [1, 1],
                    [0.4000000114992654, 1],
                    [1.149960793256355e-8, 0.5],
                    [0.5, 0.5],
                    [0.5, -0.5],
                    [-0.5, -0.5],
                    [-0.5, -0.1250000143750451],
                    [-1, -0.7500000143755803],
                    [-1, -1]
                  ]
                ],
                [
                  [
                    [-1, -0.7499999856244346],
                    [-0.5, -0.12499998562496802],
                    [-0.5, 0.5],
                    [-1.1499598773223596e-8, 0.5],
                    [0.3999999885007426, 1],
                    [-1, 1],
                    [-1, -0.7499999856244346]
                  ]
                ]
              ]
            }
          }
        ]
      },
      editType: 'split',
      featureIndexes: [0],
      editContext: null
    };

    expect(action).toEqual(expectedAction);
  });

  test('should split the multiPolygon ', () => {
    const featureCollection = {
      type: 'FeatureCollection',
      features: [createMultiPolygonFeature()]
    };
    const handler = new SplitPolygonHandler(featureCollection);
    handler.setModeConfig({});
    handler.setSelectedFeatureIndexes([0]);
    handler.handleClick(createClickEvent([-2, -2]));
    handler.handlePointerMove(createPointerMoveEvent([2, 3]));
    const action = handler.handleClick(createClickEvent([2, 3]));

    const expectedAction: EditAction = {
      updatedData: {
        type: 'FeatureCollection',
        features: [
          {
            type: 'Feature',
            properties: {},
            geometry: {
              type: 'MultiPolygon',
              coordinates: [
                [
                  [
                    [-1, -1],
                    [1, -1],
                    [1, 1],
                    [0.4000000114992654, 1],
                    [1.149960793256355e-8, 0.5],
                    [0.5, 0.5],
                    [0.5, -0.5],
                    [-0.5, -0.5],
                    [-0.5, -0.1250000143750451],
                    [-1, -0.7500000143755803],
                    [-1, -1]
                  ]
                ],
                [
                  [
                    [-1, -0.7499999856244346],
                    [-0.5, -0.12499998562496802],
                    [-0.5, 0.5],
                    [-1.1499598773223596e-8, 0.5],
                    [0.3999999885007426, 1],
                    [-1, 1],
                    [-1, -0.7499999856244346]
                  ]
                ],
                [[[2, -1], [4, -1], [4, 1], [2, 1], [2, -1]]]
              ]
            }
          }
        ]
      },
      editType: 'split',
      featureIndexes: [0],
      editContext: null
    };

    expect(action).toEqual(expectedAction);
  });
});
