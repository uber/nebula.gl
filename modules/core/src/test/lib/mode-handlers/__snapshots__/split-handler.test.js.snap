// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Split Polygon Handler  should reset tentativeFeature on handleClick  when lineString is outside the polygon lock90Degree 1`] = `
Object {
  "editContext": null,
  "editType": "split",
  "featureIndexes": Array [
    2,
  ],
  "updatedData": Object {
    "features": Array [
      Object {
        "geometry": Object {
          "coordinates": Array [
            1,
            2,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      Object {
        "geometry": Object {
          "coordinates": Array [
            Array [
              1,
              2,
            ],
            Array [
              2,
              3,
            ],
            Array [
              3,
              4,
            ],
          ],
          "type": "LineString",
        },
        "properties": Object {},
        "type": "Feature",
      },
      Object {
        "geometry": Object {
          "coordinates": Array [
            Array [
              Array [
                Array [
                  -1,
                  -1,
                ],
                Array [
                  1,
                  -1,
                ],
                Array [
                  1,
                  1,
                ],
                Array [
                  -1,
                  1,
                ],
                Array [
                  -1,
                  -0.9995433745041463,
                ],
                Array [
                  -0.5,
                  -0.49931506810778264,
                ],
                Array [
                  -0.5,
                  0.5,
                ],
                Array [
                  0.5,
                  0.5,
                ],
                Array [
                  0.5,
                  -0.5,
                ],
                Array [
                  -0.5,
                  -0.5,
                ],
                Array [
                  -0.5,
                  -0.4993150935198374,
                ],
                Array [
                  -1,
                  -0.999543399914269,
                ],
                Array [
                  -1,
                  -1,
                ],
              ],
            ],
          ],
          "type": "MultiPolygon",
        },
        "properties": Object {},
        "type": "Feature",
      },
      Object {
        "geometry": Object {
          "coordinates": Array [
            Array [
              1,
              2,
            ],
            Array [
              3,
              4,
            ],
          ],
          "type": "MultiPoint",
        },
        "properties": Object {},
        "type": "Feature",
      },
      Object {
        "geometry": Object {
          "coordinates": Array [
            Array [
              Array [
                1,
                2,
              ],
              Array [
                2,
                3,
              ],
              Array [
                3,
                4,
              ],
            ],
            Array [
              Array [
                5,
                6,
              ],
              Array [
                6,
                7,
              ],
              Array [
                7,
                8,
              ],
            ],
          ],
          "type": "MultiLineString",
        },
        "properties": Object {},
        "type": "Feature",
      },
      Object {
        "geometry": Object {
          "coordinates": Array [
            Array [
              Array [
                Array [
                  -1,
                  -1,
                ],
                Array [
                  1,
                  -1,
                ],
                Array [
                  1,
                  1,
                ],
                Array [
                  -1,
                  1,
                ],
                Array [
                  -1,
                  -1,
                ],
              ],
              Array [
                Array [
                  -0.5,
                  -0.5,
                ],
                Array [
                  -0.5,
                  0.5,
                ],
                Array [
                  0.5,
                  0.5,
                ],
                Array [
                  0.5,
                  -0.5,
                ],
                Array [
                  -0.5,
                  -0.5,
                ],
              ],
            ],
            Array [
              Array [
                Array [
                  2,
                  -1,
                ],
                Array [
                  4,
                  -1,
                ],
                Array [
                  4,
                  1,
                ],
                Array [
                  2,
                  1,
                ],
                Array [
                  2,
                  -1,
                ],
              ],
            ],
          ],
          "type": "MultiPolygon",
        },
        "properties": Object {},
        "type": "Feature",
      },
    ],
    "type": "FeatureCollection",
  },
}
`;
